rdigest.1
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 06/06/2014               \" DATE 
.Dt rdigest 1      \" Program name and manual section number 
.Sh NAME                 \" Section Header - required - don't modify 
.Nm rdigest
.Nd calculates digest of files recursively found under directories.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl qhbv
.Op Fl o Ar outfile
fileOrDirectory ...
.Sh DESCRIPTION          \" Section Header - required - don't modify
Recursively examines the contents of directories and generates a summary report
of their contents. The report can contains the SHA-1 digest of the contents of
the files, so that changes can be detected.
.Pp
This program is useful for comparing sets of files, to see if they are
the same or not. If the output of the script is different then the two
sets of files are different.
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl q | Fl Fl quick  \"-a flag as a list item
Produce a report using file sizes instead of digests.  This is much
faster, because the contents of the files are not examined. But the
report cannot be used to detect changes to the file's contents if the
change does not affect the file's size.
.It Fl b | Fl Fl baseless
Do not include pathnames with the specified directory in the results.
.It Fl o | Fl Fl output Ar outfile
Write report to given output file. Default is to
.Em stdout .
.It Fl v | Fl Fl verbose
Output additional statistics at the end of the run.
.It Fl h | Fl Fl help
Show a brief summary of the options.
.El                      \" Ends the list
.Sh REPORT FORMAT
The report consists one entry per line. Each entry consists of the
entry type followed by a left parenthesis, the file name, right
parenthesis and the value. The value consists of an equal sign, space,
and the encoded value.
.Pp
For digest entries, the entry type is "SHA1" and the encoded value is
the SHA-1 digest value represented in hexadecimal.
.Pp
If the quick mode is used, size entries are produced instead of digest
entries.  Size entries have the entry type of "SIZE" and the encoded
value is the number of bytes in decimal.
.Pp
Symbolic links appear as "SYMLINK" entries, where the encoded value is
the path it links to.
.Pp
Empty directories appear as "EMPTY_DIRECTORY" entries, without any
value. Non-empty directories do not appear as a separate entries,
since their existance is implicitly represented by the entries for the
files and/or directories under it.
.Sh EXAMPLES

rdigest dir1 --output report1.sha1
rdigest dir2 --output report2.sha1
diff report1.sha1 report2.sha1

rdigest dir1 --quick --output report1.size

.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "/usr/local/bin/rdigest" -compact
.It Pa /usr/local/bin/rdigest
Executable program
.El                      \" Ends the list
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr openssl 1 , 
.Xr dgst 1 ,
.Sh BUGS              \" Document known, unremedied bugs 
Only ordinary files, directories and symbolic links are handled. Other
types of files (e.g. sockets, FIFOs, character and block devices) are
not expected under normal use and are not handled.
.Pp
Only the file name and file contents (or just its size, if quick mode
is used) is used. Other file properties (such as dates, permissions
and extended attributes) are not used in the report.
.Sh HISTORY           \" Document history if command behaves in a unique manner
The digest entry is the same as the output produced by the 
.Em openssl dgst -sha1
OpenSSL digest command.  This program was originally written to
generate a report for all the files under a directory, because the
OpenSSL digest command could not recursively scan a directory for
files and there was a limit to the number of command line arguments.


rdigest.pl
#!/usr/bin/perl

use strict;
use warnings;

use File::Basename;
use File::Find;
use Getopt::Long;

#----------------------------------------------------------------

my $VERSION = '1.1';

my $PROG = basename($0);

# Globals

my $quick_mode = 0;
my $combined_mode = 0;

my $combined_size = undef; # only used if (combined_mode && quick_mode)
my $combined_sha1 = undef; # only used if (combined_mode && ! quick_mode)

my $num_files = 0;

# Detect SHA1 calculator to use

# The Perl module is preferred since it is much faster than invoking
# openssl as an separate process. If neither is available, this
# program can still perform the non-digest operations. The variable
# $DIGEST_CALCULATOR is either undef (no digest calculator available),
# the value of $DIGEST_CALCULATOR_PERL_MODULE (when Digest::SHA1 is
# used), or any other string value (when OpenSSL is used).

my $DIGEST_CALCULATOR_PERL_MODULE = 'Digest::SHA1';

my $DIGEST_CALCULATOR;
eval { require Digest::SHA1; };
if (! $@) {
  $DIGEST_CALCULATOR = $DIGEST_CALCULATOR_PERL_MODULE; # Perl module is used
} else {
  my $openssl_ver = `openssl version 2>/dev/null`;
  if (defined($openssl_ver) && $openssl_ver ne '') {
    $openssl_ver =~ s/\s+$//;
    $DIGEST_CALCULATOR = $openssl_ver; # OpenSSL is used
  } else {
    $DIGEST_CALCULATOR = undef; # no digest calculator
  }
}

#----------------------------------------------------------------

sub process_file {
  my($filename) = @_;

  if ($quick_mode) {
    # Use size of file

    my $file_size = (stat($filename))[7]; # size
    if ($combined_mode) {
      $combined_size += $file_size;
    } else {
      print 'SIZE(' . $File::Find::name . ')= ' . $file_size . "\n";
    }

  } else {
    # Use SHA1 digest of file contents

    my $str;
    if (defined($DIGEST_CALCULATOR) &&
	$DIGEST_CALCULATOR eq $DIGEST_CALCULATOR_PERL_MODULE) {
      # Use Perl module
      open(FILE, '<', $filename) || die "Error: $!: $filename\n";
      binmode FILE;
      my $sha1 = Digest::SHA1->new;
      $sha1->addfile(*FILE);
      $str = 'SHA1(' . $File::Find::name . ')= ' . $sha1->hexdigest . "\n";
      close(FILE);
    } else {
      $str = `openssl dgst -sha1 '$filename'`;
    }

    if ($combined_mode) {
      if (defined($DIGEST_CALCULATOR) &&
	  $DIGEST_CALCULATOR eq $DIGEST_CALCULATOR_PERL_MODULE) {
        $combined_sha1->add($str);
      } else {
        print COMBINE $str;
      }
    } else {
      print $str;
    }

  }

  $num_files++;
}

#----------------------------------------------------------------

sub process_arguments {
  my ($combined_ref, $quick_ref, $output_ref, $verbose_ref) = @_;
  my $help;

  if (! GetOptions('verbose' => $verbose_ref,
		   'combined' => $combined_ref,
		   'quick' => $quick_ref,
		   'output=s' => $output_ref,
		   'help' => \$help)) {
    exit(1);
  }

  if ($help) {
    print "Usage: $PROG [options] {dirOrFile...}\n";
    print "Options:\n";
    print "  --help        show this help message\n";
    print "  --output dir  file to write digests to\n";
    print "  --combined    calculate a single combined value for all files\n";
    print "  --quick       use size of files instead of calculating digests\n";
    print "  --verbose     print number of files processed at end\n";
    print "Version: $VERSION\n";
    print "Digest calculator: ";
    if (defined($DIGEST_CALCULATOR)) {
      print "$DIGEST_CALCULATOR\n";
    } else {
      print "none (digest calculating operations are not available)\n";
    }
    exit(0);
  }

  # Arguments

  if (scalar(@ARGV) < 1) {
    die "Usage error: no directories or files specified (-h for help)\n";
  }
}

#----------------------------------------------------------------

sub main {
  my $output;
  my $verbose;
  process_arguments(\$combined_mode, \$quick_mode, \$output, \$verbose);

  # Check if all items to process exist (better to fail fast)

  my $error = 0;
  foreach my $item (@ARGV) {
    if (! -e $item) {
      print STDERR "Error: file or directory does not exist: $item\n";
      $error++;
    }
  }
  if ($error) {
    exit(1);
  }

  if (defined($output)) {
    # Redirect stdout to file
    open(STDOUT, '>', $output) || die "Error: $!: $output\n";
  }

  # Prepare for processing

  if ($combined_mode) {
    if ($quick_mode) {
      $combined_size = 0;
    } else {
      if (defined($DIGEST_CALCULATOR)) {
	if ($DIGEST_CALCULATOR eq $DIGEST_CALCULATOR_PERL_MODULE) {
	  $combined_sha1 = Digest::SHA1->new;
	} else {
	  open(COMBINE, "|openssl dgst -sha1 | sed -e 's/(stdin)\= //'") || 
	      die "Error: openssl: $!\n";
	}
      } else {
	die "Internal error: no digest calculator: operation not supported\n";
      }
    }
  } else {
    if (! $quick_mode && ! defined($DIGEST_CALCULATOR)) {
      die "Internal error: no digest calculator: operation not supported\n";
    }
  }
  $num_files = 0;

  # Process all files

  foreach my $item (@ARGV) {
    find( { wanted => sub { process_file($_) if (-f $_) },
	    no_chdir => 1 }, $item);
  }

  # Output summary results

  if ($combined_mode) {
    if ($quick_mode) {
      print $combined_size, (($combined_size == 1) ? ' byte' : ' bytes'), "\n";
    } else {
      if (defined($DIGEST_CALCULATOR) &&
	  $DIGEST_CALCULATOR eq $DIGEST_CALCULATOR_PERL_MODULE) {
        print $combined_sha1->hexdigest, "\n";
      } else {
        close(COMBINE);
      }
    }
  }

  if ($verbose) {
    print $num_files, (($num_files == 1) ? ' file' : ' files'), "\n";
  }

  return 0;
}

exit main();

#EOF
